File: app/src/main/java/me/pwcong/simplesidebar/view/SimpleSideBar.java
package me.pwcong.simplesidebar.view;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Typeface;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import me.pwcong.simplesidebar.R;

/**
 * Created by pwcong on 2016/8/2.
 */
public class SimpleSideBar extends View {

    OnLetterTouchedChangeListener onLetterTouchedChangeListener;

    public void setOnLetterTouchedChangeListener(OnLetterTouchedChangeListener onLetterTouchedChangeListener) {
        this.onLetterTouchedChangeListener = onLetterTouchedChangeListener;
    }

    // 索引字母数组
    private String[] alphabet = {
            "A", "B", "C", "D", "E", "F",
            "G", "H", "I", "J", "K", "L",
            "M", "N", "O", "P", "Q", "R",
            "S", "T", "U", "V", "W", "X",
            "Y", "Z"
    };

    // 当前选择的索引字母的下标
    private int currentChoosenAlphabetIndex=-1;

    private Paint mPaint=new Paint();

    private int alphabetTextSize=20;

    public SimpleSideBar(Context context) {
        super(context);
    }

    public SimpleSideBar(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public SimpleSideBar(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }


    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        int viewHeight=getHeight();
        int viewWidth=getWidth();
        int heightPerAlphabet=viewHeight/alphabet.length;

        for (int i=0;i<alphabet.length;i++){

            mPaint.setColor(Color.BLACK);
            mPaint.setTypeface(Typeface.DEFAULT_BOLD);
            mPaint.setTextSize(alphabetTextSize);
            mPaint.setAntiAlias(true);

            if(currentChoosenAlphabetIndex==i){

                mPaint.setColor(Color.YELLOW);
                mPaint.setTextSize(alphabetTextSize);
                mPaint.setFakeBoldText(true);

            }

            float xPos=viewWidth/2-mPaint.measureText(alphabet[i])/2;
            float yPos=heightPerAlphabet*i+heightPerAlphabet;
            canvas.drawText(alphabet[i],xPos,yPos,mPaint);
            mPaint.reset();

        }
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent event) {

        int action = event.getAction();

        float touchYPos=event.getY();


        int currentTouchIndex= (int) (touchYPos/getHeight()*alphabet.length);


        switch (action){

            case MotionEvent.ACTION_DOWN:
            case MotionEvent.ACTION_MOVE:
                setBackgroundResource(R.color.grey_600);
                currentChoosenAlphabetIndex=currentTouchIndex;

                // 如果接口
                if(onLetterTouchedChangeListener!=null&&currentTouchIndex<alphabet.length&&currentTouchIndex>-1){
                    onLetterTouchedChangeListener.onTouchedLetterChange(alphabet[currentTouchIndex]);
                }

                invalidate();


                break;
            case MotionEvent.ACTION_UP:
                setBackgroundResource(R.color.grey_50);
                currentChoosenAlphabetIndex=-1;
                invalidate();
                break;
            default:break;

        }

        return true;
    }

    public interface OnLetterTouchedChangeListener{

        void onTouchedLetterChange(String letterTouched);

    }

}

File: simplesidebar/src/main/java/me/pwcong/simplesidebar/view/SimpleSideBar.kt
package me.pwcong.simplesidebar.view

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Typeface
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View

import me.pwcong.simplesidebar.R

class SimpleSideBar : View {
    private var onLetterTouchedChangeListener: OnLetterTouchedChangeListener? = null

    fun setOnLetterTouchedChangeListener(onLetterTouchedChangeListener: OnLetterTouchedChangeListener?) {
        this.onLetterTouchedChangeListener = onLetterTouchedChangeListener
    }

    // 索引字母数组
    private val alphabet = arrayOf(
        "A", "B", "C", "D", "E", "F",
        "G", "H", "I", "J", "K", "L",
        "M", "N", "O", "P", "Q", "R",
        "S", "T", "U", "V", "W", "X",
        "Y", "Z"
    )

    // 当前选择的索引字母的下标
    private var currentChoosenAlphabetIndex = -1

    private val mPaint = Paint()

    private val alphabetTextSize = 20

    constructor(context: Context?) : super(context)

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs)

    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    )


    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        val viewHeight = height
        val viewWidth = width
        val heightPerAlphabet = viewHeight / alphabet.size

        for (i in alphabet.indices) {
            mPaint.color = Color.BLACK
            mPaint.setTypeface(Typeface.DEFAULT_BOLD)
            mPaint.textSize = alphabetTextSize.toFloat()
            mPaint.isAntiAlias = true

            if (currentChoosenAlphabetIndex == i) {
                mPaint.color = Color.YELLOW
                mPaint.textSize = alphabetTextSize.toFloat()
                mPaint.isFakeBoldText = true
            }

            val xPos = viewWidth / 2 - mPaint.measureText(alphabet[i]) / 2
            val yPos = (heightPerAlphabet * i + heightPerAlphabet).toFloat()
            canvas.drawText(alphabet[i], xPos, yPos, mPaint)
            mPaint.reset()
        }
    }

    override fun dispatchTouchEvent(event: MotionEvent): Boolean {
        val action = event.action

        val touchYPos = event.y


        val currentTouchIndex = (touchYPos / height * alphabet.size).toInt()


        when (action) {
            MotionEvent.ACTION_DOWN, MotionEvent.ACTION_MOVE -> {
                setBackgroundResource(R.color.grey_active)
                currentChoosenAlphabetIndex = currentTouchIndex

                // 如果接口
                if (onLetterTouchedChangeListener != null && currentTouchIndex < alphabet.size && currentTouchIndex > -1) {
                    onLetterTouchedChangeListener!!.onTouchedLetterChange(alphabet[currentTouchIndex])
                }

                invalidate()
            }

            MotionEvent.ACTION_UP -> {
                setBackgroundResource(R.color.grey)
                currentChoosenAlphabetIndex = -1
                invalidate()
            }

            else -> {}
        }
        return true
    }

    interface OnLetterTouchedChangeListener {
        fun onTouchedLetterChange(letterTouched: String?)
    }
}
